python/cudf/cudf/core/udf/function.cu

```diff
@@ -0,0 +1,817 @@
+    while(val < __longlong_as_double(old))
+    {
+        unsigned long long assumed = old;
+        if((old = atomicCAS((unsigned long long *)address, assumed, __double_as_longlong(val))) == assumed)
+            break;
+    }
+    return __longlong_as_double(old);
+}
+
+extern "C" __device__ int BlockSum_int64(int64_t *numba_return_value, int64_t *data, int64_t size) {
+  int tid = threadIdx.x; int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  int ITEMS_PER_THREAD = (size + tb_size - 1) / tb_size;
+  int64_t local_sum = 0;
+
+  __shared__ int64_t sum;
+
+  if (tid == 0)
+    sum = 0;
+
+  __syncthreads();
+
+  // Calculate local sum for each thread
+  #pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+      if (tid + (ITEM * tb_size) < size) {
+          int64_t load = data[tid + ITEM * tb_size];
+          local_sum += load;
+      }
+  }
+
+  atomicAdd((unsigned long long*) &sum, (unsigned long long) local_sum);
+
+  __syncthreads();
+
+  *numba_return_value = sum;
+
+  return 0;
+}
+
+extern "C" __device__ int BlockSum_float64(double *numba_return_value, double *data, int64_t size) {
+  int tid = threadIdx.x; int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  int ITEMS_PER_THREAD = (size + tb_size - 1) / tb_size;
+  double local_sum = 0;
+
+  __shared__ double sum;
+
+  if (tid == 0)
+    sum = 0;
+
+  __syncthreads();
+
+  // Calculate local sum for each thread
+  #pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+      if (tid + (ITEM * tb_size) < size) {
+          double load = data[tid + ITEM * tb_size];
+          local_sum += load;
+      }
+  }
+
+  atomicAdd(&sum, local_sum);
+
+  __syncthreads();
+
+  *numba_return_value = sum;
+
+  return 0;
+}
+
+
+extern "C" __device__ int BlockMean_int64(double *numba_return_value, int64_t *data, int64_t size) {
+  int tid = threadIdx.x; int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  int ITEMS_PER_THREAD = (size + tb_size - 1) / tb_size;
+  int64_t local_sum = 0;
+  double mean;
+
+  __shared__ int64_t sum;
+
+  if (tid == 0)
+    sum = 0;
+
+  __syncthreads();
+
+  // Calculate local sum for each thread
+  #pragma unroll
+  for (int ITEM = 0; ITEM < ITEMS_PER_THREAD; ITEM++) {
+      if (tid + (ITEM * tb_size) < size) {
+          int64_t load = data[tid + ITEM * tb_size];
+          local_sum += load;
+      }
+  }
+
+  atomicAdd((unsigned long long*) &sum, (unsigned long long) local_sum);
+
+  __syncthreads();
+
+  mean = sum * 1.0 / size;
+
+  *numba_return_value = mean;
+
+  return 0;
+
+}
+
+extern "C" __device__ int BlockMean_float64(double *numba_return_value, double *data, int64_t size) {
+  int tid = threadIdx.x; int tb_size = blockDim.x;
+  // Calculate how many elements each thread is working on
+  int ITEMS_PER_THREAD = (size + tb_size - 1) / tb_size;
+  double local_sum = 0;
+  double mean;
+
+  __shared__ double sum;
```

Comment:
It's a little concerning to me that we have this entire function repeated so many times. It appears to need C linkage and thus you can't use templates directly:

> ... A template, a template explicit specialization (17.7.3), and a class template partial specialization shall not have C linkage. ...
https://stackoverflow.com/a/52367443

However, you should be able to work around this creatively, and significantly reduce the lines of code here. I have a proposal: https://godbolt.org/z/ondMPxsjc

The core of what I'm proposing is this:
```c++
// Use a C++ templated __device__ function to implement the body of the algorithm.
template <typename T>
double sum(T const* values, size_type num_values) {
    double total = 0;
    for (size_type i = 0; i < num_values; i++) {
        total += values[i];
    }
    return total;
}

// Then use that C++ templated __device__ function in each of the extern "C" functions.
extern "C" int BlockSum_int64(int64_t *numba_return_value, int64_t *data, int64_t size) {
  *numba_return_value = sum(data, size);
  return 0;
}

extern "C" int BlockSum_float64(double *numba_return_value, double *data, int64_t size) {
  *numba_return_value = sum(data, size);
  return 0;
}
```

Reply:
The other HUGE benefit to this is that you can use these functions as building blocks -- you don't need to write `mean` as a standalone function, you can write `mean` with a call to `sum` and division by the size.

tl;dr This file can be _much_ shorter if you make the code composable and reusable. It'll all compile out the same.

Reply:
Templating outside of C++, e.g., via something like `jinja` is possible too. In particular, [PyCUDA uses this trick](https://documen.tician.de/pycuda/metaprog.html) a lot to generate CUDA kernels for different type combinations.

Reply:
Should we be using CUB for these device-side reduction operations, rather than hand writing each one (even if only once)?

Reply:
> Should we be using CUB for these device-side reduction operations, rather than hand writing each one

Absolutely yes. The first step would be removing duplicates by using template first then thinking about how to replace the custom kernels with corresponding CUB functions.

Reply:
That’s exactly my plan of attack here, too. 😄 The implementation in CUB will be the most ideal for this. We shouldn’t be hardcoding most (any?) of these algorithms, but deduplication is where I’d start to centralize the problem.

CUB wasn’t my first suggestion because I believe it will require the JIT compiler to know more information, like the paths to the proper include directory and also GPU arch information— but I think this is shipped as a raw PTX output, and I didn’t want to think about how CUB interacts with GPU arch version compatibility yet.